upup/pkg/fi/cloudup/dns.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudup
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	api "k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/featureflag"
+	"k8s.io/kops/pkg/model"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+	"net"
+	"os"
+	"strings"
+)
+
+const (
+	// This IP is from TEST-NET-3
+	// https://en.wikipedia.org/wiki/Reserved_IP_addresses
+	PlaceholderIP  = "203.0.113.123"
+	PlaceholderTTL = 10
+)
+
+func findZone(cluster *api.Cluster, cloud fi.Cloud) (dnsprovider.Zone, error) {
+	dns, err := cloud.DNS()
+	if err != nil {
+		return nil, fmt.Errorf("error building DNS provider: %v", err)
+	}
+
+	zonesProvider, ok := dns.Zones()
+	if !ok {
+		return nil, fmt.Errorf("error getting DNS zones provider")
+	}
+
+	zones, err := zonesProvider.List()
+	if err != nil {
+		return nil, fmt.Errorf("error listing DNS zones: %v", err)
+	}
+
+	var matches []dnsprovider.Zone
+	findName := strings.TrimSuffix(cluster.Spec.DNSZone, ".")
+	for _, zone := range zones {
+		id := zone.ID()
+		name := strings.TrimSuffix(zone.Name(), ".")
+		if id == cluster.Spec.DNSZone || name == findName {
+			matches = append(matches, zone)
+		}
+	}
+	if len(matches) == 0 {
+		return nil, fmt.Errorf("cannot find DNS Zone %q.  Please pre-create the zone and set up NS records so that it resolves.", cluster.Spec.DNSZone)
+	}
+
+	if len(matches) > 1 {
+		return nil, fmt.Errorf("found multiple DNS Zones matching %q", cluster.Spec.DNSZone)
+	}
+
+	zone := matches[0]
+	return zone, nil
+}
+
+func validateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
+	kopsModelContext := &model.KopsModelContext{
+		//Region: cloud.Region(),
```

Comment:
Added a comment to better clarify the point I was trying to make - that it isn't a "full" KopsModelContext